rfcs/20190115-dense-attention.md

```diff
@@ -0,0 +1,532 @@
+# Attention for Dense networks on Keras
+
+| Status        | Proposed                                                                   |
+:-------------- |:-------------------------------------------------------------------------- |
+| **Author(s)** | Georgios Roumpos (roumposg@google.com)                                     |
+| **Sponsors**  | Karmel Allison (karmel@google.com), Francois Chollet (fchollet@google.com) |
+| **Updated**   | 2019-01-15                                                                 |
+
+## Objective and Motivation
+
+Recently people have had success using the Attention mechanism in dense layers,
+e.g. CNN+Attention or Transformer networks. Some examples are the
+["Attention is all you need"](https://arxiv.org/abs/1706.03762) paper, and
+models in
+[semantic text similarity](https://ai.googleblog.com/2018/05/advances-in-semantic-textual-similarity.html).
+`tf.keras.layers` is the recommended way to build models in tensorflow, but it
+does not have a layer for attention that works with CNN/DNN networks. We would
+like to contribute this capability.
+
+Keras is an API spec that can be implemented across different languages and
+backends, and `tf.keras` is a particular implementation of that spec. This
+document contains code examples for `tensorflow`, but the same API should work
+everywhere.
+
+### Recurrent Neural Networks
+
+Although not the primary focus of this proposal, the Attention layers in this
+proposal work with some configurations of RNN networks.
+Namely, when users create
+[tf.keras.layers.LSTM](https://www.tensorflow.org/api_docs/python/tf/keras/layers/LSTM)
+with `return_sequences=True`, the rest works the same way as CNN.
+
+Unfortunately, this technique does not cover sequence-to-sequence RNN models.
+In these models, the value is the states of encoder, and the query is the input
+of the decoder. The decoder needs to slide its input based on the timesteps, and
+feed them one by one. So, the output of the attention layer at timestep T
+affects the output at T+1.
+
+## Previous Work
+
+[tf.contrib.seq2seq.LuongAttention](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/LuongAttention)
+and
+[tf.contrib.seq2seq.BahdanauAttention](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BahdanauAttention)
+are implementations of dot-product (Luong) and additive (Bahdanau) Attention
+respectively for RNN in Tensorflow. This proposal is based on this
```

Comment:
By "based", do you mean it will extend classes from `tf.seq2seq`? It's unclear to me if this proposal has this recent change in mind: https://github.com/tensorflow/tensorflow/commit/b797012b39f89c45f13336b36c6b73e9ad93d815.

Reply:
Thank you for you comment. I mean that it is inspired from it, and that the implementation is consistent. Let me update the text.

Yes, I am aware of this change. It covers RNN networks, and it moves in the same direction as my proposal. Namely, implementing as keras.layer. Let me update the text to clarify that.